misc/incremental_checker.py

```diff
@@ -0,0 +1,218 @@
+    if stderr != "":
+        output = stderr
+    delta = time.time() - start
+    return delta, output
+
+
+def load_cache(cache_path: str = CACHE_PATH) -> JsonDict:
+    if os.path.exists(cache_path):
+        with open(cache_path, 'r') as stream:
+            return json.load(stream)
+    else:
+        return {}
+
+
+def save_cache(cache: JsonDict, cache_path: str = CACHE_PATH) -> None:
+    with open(cache_path, 'w') as stream:
+        json.dump(cache, stream, indent=2)
+
+
+def set_expected(commits: List[Tuple[str, str]],
+                 cache: JsonDict,
+                 repo_name: str,
+                 target: str) -> None:
+    for commit_id, message in commits:
+        if commit_id in cache:
+            print('Skipping commit (already cached): {0}: "{1}"'.format(commit_id, message))
+        else:
+            print('Caching expected output for commit {0}: "{1}"'.format(commit_id, message))
+            execute(["git", "-C", repo_name, "checkout", commit_id])
+            delta, output = run_mypy(target, incremental=False)
+            cache[commit_id] = {'delta': delta, 'output': output}
+            if output == "":
+                print("    Clean output ({:.3f} sec)".format(delta))
+            else:
+                print("    Output ({:.3f} sec)".format(delta))
+                print_offset(output, 8)
+    print()
+
+
+def test_incremental(commits: List[Tuple[str, str]],
+                     cache: JsonDict,
+                     repo_name,
+                     target: str) -> None:
+    print("Note: first commit is evaluated twice to warm up cache")
+    commits = [commits[0]] + commits
+    for commit_id, message in commits:
+        print('Now testing commit {0}: "{1}"'.format(commit_id, message))
+        execute(["git", "-C", repo_name, "checkout", commit_id])
+        delta, output = run_mypy(target)
+        expected_delta = cache[commit_id]['delta']  # type: float
+        expected_output = cache[commit_id]['output']  # type: str
+        if output != expected_output:
+            print("    Output does not match expected result!")
+            print("    Expected output ({:.3f} sec):".format(expected_delta))
+            print_offset(expected_output, 8)
+            print("    Actual output: ({:.3f} sec):".format(delta))
+            print_offset(output, 8)
+        else:
+            print("    Output matches expected result!")
+            print("    Incremental: {:.3f} sec".format(delta))
+            print("    Original:    {:.3f} sec".format(expected_delta))
+
+
+def cleanup(folder_path: str) -> None:
+    print("Cleanup: deleting {}".format(folder_path))
+    delete_folder(folder_path)
+    delete_folder("misc/.mypy_cache")
+
+
+def test_repo(target_repo_url: str, module_name: str, start_commit: str,
+              branch: str, cache_path: str) -> None:
+    repo_folder_name = "tmp_repo"
+    target = os.path.join(repo_folder_name, module_name)
+
+    # Assume we're starting in misc
```

Comment:
Hm... I ran it from the repo root using `python3 misc/incremental_checker.py` -- maybe it shouldn't matter from where it's ran?